tsdb/index/index.go

```diff
@@ -921,25 +960,82 @@ func (r *Reader) Series(id uint64, lbls *labels.Labels, chks *[]chunks.Meta) err
 	return errors.Wrap(r.dec.Series(d.Get(), lbls, chks), "read series")
 }
 
-// Postings returns a postings list for the given label pair.
-func (r *Reader) Postings(name, value string) (Postings, error) {
+func (r *Reader) Postings(name string, values ...string) (Postings, error) {
 	e, ok := r.postings[name]
 	if !ok {
 		return EmptyPostings(), nil
 	}
-	off, ok := e[value]
-	if !ok {
+
+	if len(values) == 0 {
 		return EmptyPostings(), nil
 	}
-	d := encoding.NewDecbufAt(r.b, int(off), castagnoliTable)
-	if d.Err() != nil {
-		return nil, errors.Wrap(d.Err(), "get postings entry")
+
+	res := make([]Postings, 0, len(values))
+	skip := 0
+	valueIndex := 0
+	for valueIndex < len(values) && values[valueIndex] < e[0].value {
+		// Discard values before the start.
+		valueIndex++
 	}
-	_, p, err := r.dec.Postings(d.Get())
-	if err != nil {
-		return nil, errors.Wrap(err, "decode postings")
+	for valueIndex < len(values) {
+		value := values[valueIndex]
+
+		i := sort.Search(len(e), func(i int) bool { return e[i].value >= value })
```

Comment:
If I'm not missing anything, postings values are sorted. Given this function picks multiple values in input (and they're sorted as well), the `values[i+1]` is > `values[i]`, which means the position of `values[i+1]` within `e` will be at a position > than the previous one, which means we may reduce the iterations done by `sort.Search()` with something like (pseudo-code):

```
i := sort.Search(len(e) - previousPos, func(i int) bool { return e[i+previousPos].value >= value })
i += previousPos
```

Reply:
Yes that's an optimisation I considered, however the code is pretty finicky so I left it out for now.